contrib/cloud.google.com/go/pubsub.v1/pubsub_test.go

```diff
@@ -0,0 +1,218 @@
+
+package pubsub
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"cloud.google.com/go/pubsub"
+	"cloud.google.com/go/pubsub/pstest"
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/api/option"
+	"google.golang.org/grpc"
+)
+
+func TestPropagation(t *testing.T) {
+	ctx, topic, sub, mt, cleanup := setup(t)
+	defer cleanup()
+
+	// Publisher
+	span, pctx := tracer.StartSpanFromContext(ctx, "propagation-test", tracer.WithSpanID(42)) // set the root trace ID
+	_, err := Publish(pctx, topic, &pubsub.Message{Data: []byte("hello"), OrderingKey: "xxx"})
+	assert.NoError(t, err)
+	span.Finish()
+
+	// Subscriber
+	var (
+		msgID  string
+		spanID uint64
+		called bool
+	)
+	err = sub.Receive(ctx, ReceiveTracer(sub, func(ctx context.Context, msg *pubsub.Message) {
+		assert.False(t, called, "callback called twice")
+		assert.Equal(t, msg.Data, []byte("hello"), "wrong payload")
+		span, ok := tracer.SpanFromContext(ctx)
+		assert.True(t, ok, "no span")
+		assert.Equal(t, uint64(42), span.Context().TraceID(), "wrong trace id") // gist of the test: the trace ID must be the same as the root trace ID set above
+		msgID = msg.ID
+		spanID = span.Context().SpanID()
+		msg.Ack()
+		called = true
+	}))
+	assert.True(t, called, "callback not called")
+	assert.NoError(t, err)
+
+	spans := mt.FinishedSpans()
+	assert.Len(t, spans, 3, "wrong number of spans")
+	assert.Equal(t, "pubsub.publish", spans[0].OperationName())
+	assert.Equal(t, "propagation-test", spans[1].OperationName())
+	assert.Equal(t, "pubsub.receive", spans[2].OperationName())
+
+	assert.Equal(t, ext.SpanTypeMessageProducer, spans[0].Tag(ext.SpanType))
+	assert.Equal(t, "projects/project/topics/topic", spans[0].Tag(ext.ResourceName))
+	assert.Equal(t, spans[1].SpanID(), spans[0].ParentID())
+	assert.Equal(t, uint64(42), spans[0].TraceID())
+	assert.Equal(t, 5, spans[0].Tag("message_size"))
+	assert.Equal(t, 2, spans[0].Tag("num_attributes")) // 2 tracing attributes
+	assert.Equal(t, "xxx", spans[0].Tag("ordering_key"))
+	assert.Empty(t, spans[0].Tag("message_id"))
+	assert.Empty(t, spans[0].Tag("publish_time"))
+	assert.Empty(t, spans[0].Tag(ext.Error))
+
+	assert.Equal(t, ext.SpanTypeMessageConsumer, spans[2].Tag(ext.SpanType))
+	assert.Equal(t, "projects/project/subscriptions/subscription", spans[2].Tag(ext.ResourceName))
+	assert.Equal(t, spans[0].SpanID(), spans[2].ParentID())
+	assert.Equal(t, uint64(42), spans[2].TraceID())
+	assert.Equal(t, spanID, spans[2].SpanID())
+	assert.Equal(t, 5, spans[2].Tag("message_size"))
+	assert.Equal(t, 2, spans[2].Tag("num_attributes"))
+	assert.Equal(t, "xxx", spans[2].Tag("ordering_key"))
+	assert.NotEmpty(t, spans[2].Tag("message_id"))
+	assert.Equal(t, msgID, spans[2].Tag("message_id"))
+	assert.NotEmpty(t, spans[2].Tag("publish_time"))
+	assert.Empty(t, spans[2].Tag(ext.Error))
```

Comment:
We sometimes do:
```go
assert := assert.New(t)
```
It's a weird habit but it'll avoid the first `t` argument everywhere.